polynote-frontend/polynote/ui.js

```diff
@@ -918,32 +927,45 @@ export class NotebookUI {
             }
         });
 
-        socket.addMessageListener(messages.UpdateCell, (path, globalVersion, localVersion, id, edits) => {
-            if (globalVersion >= this.globalVersion) {
-                this.globalVersion = globalVersion;
-                let rebasedEdits = edits;
+        socket.addMessageListener(messages.NotebookUpdate, update => {
+            if (update.path === this.path) {
+                if (update.globalVersion >= this.globalVersion) {
+                    this.globalVersion = update.globalVersion;
 
-                if (localVersion < this.localVersion) {
-                    const editsBefore = this.editBuffer.range(localVersion, this.localVersion);
-                    rebasedEdits = messages.ContentEdit.rebaseEdits(edits, editsBefore);
-                } else if (localVersion > this.localVersion) {
-                    console.error("Local version from server is after client's. Something's wrong.")
-                }
+                    if (update.localVersion < this.localVersion) {
+                        const prevUpdates = this.editBuffer.range(update.localVersion, this.localVersion);
+                        update = messages.NotebookUpdate.rebase(update, prevUpdates);
+                    }
 
-                this.localVersion++;
 
-                const cell = this.cellUI.getCell(id);
+                    this.localVersion++;
+
+                    match(update)
```

Comment:
How I wish JS had `_` for ignored things (it's actually just an identifier, so you could use it, but only once! D'oh!)